File: Delphi/DBComparerDemoADO/Unit1Ado.pas
unit Unit1Ado;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, Buttons, StdCtrls,
  ExtCtrls, dbcMSSQLDatabaseExtract, dbcClasses, dbcCustomScriptExtract,
  dbcMSSQLScriptExtract, dbcSQL_Exec, dbcMSSQLExec, dbcDBEngine,
  dbcConnection_ADO, dbcDBComparer, dbcDBStructure, ComCtrls;

type
  TForm1 = class(TForm)
    DBStructure1: TDBStructure;
    DBStructure2: TDBStructure;
    OpenDialog1: TOpenDialog;
    GroupBox1: TGroupBox;
    EditDbMaster: TEdit;
    EditDbTarget: TEdit;
    Label1: TLabel;
    Label2: TLabel;
    BtnSelDatabaseMaster: TSpeedButton;
    BtnSelDatabaseTarget: TSpeedButton;
    EditMUser: TEdit;
    EditMPswd: TEdit;
    LabelMUser: TLabel;
    LabelMPswd: TLabel;
    LabelTUser: TLabel;
    EditTUser: TEdit;
    LabelTPswd: TLabel;
    EditTPswd: TEdit;
    BtnExtract: TBitBtn;
    BtnCompare: TBitBtn;
    GroupBox2: TGroupBox;
    MemoLog: TRichEdit;
    DBComparer1: TDBComparer;
    Splitter1: TSplitter;
    GroupBox3: TGroupBox;
    MemoResult: TMemo;
    fnScriptMaster: TEdit;
    BtnSelScriptMaster: TSpeedButton;
    fnScriptTarget: TEdit;
    BtnSelScriptTarget: TSpeedButton;
    Panel1: TPanel;
    rbMasterFromDatabase: TRadioButton;
    rbMasterFromScript: TRadioButton;
    Panel2: TPanel;
    rbTargetFromDatabase: TRadioButton;
    rbTargetFromScript: TRadioButton;
    BtnUpdate: TBitBtn;
    Gauge1: TProgressBar;
    DBCConnectionADO1: TDBCConnectionADO;
    DBCConnectionADO2: TDBCConnectionADO;
    MSSQLExec1: TMSSQLExec;
    MSSQLScriptExtract1: TMSSQLScriptExtract;
    MSSQLScriptExtract2: TMSSQLScriptExtract;
    MSSQLDBExtract1: TMSSQLDBExtract;
    MSSQLDBExtract2: TMSSQLDBExtract;
    Panel3: TPanel;
    Panel4: TPanel;
    LabelSrvM: TLabel;
    EditHostM: TEdit;
    LabelAuthM: TLabel;
    rbMWin: TRadioButton;
    rbMSql: TRadioButton;
    LabelSrvT: TLabel;
    EditHostT: TEdit;
    LabelAuthT: TLabel;
    rbTWin: TRadioButton;
    rbTSql: TRadioButton;
    procedure FormCreate(Sender: TObject);
    procedure BtnSelFileClick(Sender: TObject);
    procedure BtnExtractClick(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure LogNextLine(Sender: TObject; LogLine: String);
    procedure BtnCompareClick(Sender: TObject);
    procedure rbFromClick(Sender: TObject);
    procedure ProgressUpdate(Sender: TObject; Value, MaxValue: Integer);
    procedure BtnUpdateClick(Sender: TObject);
  private
    { Private declarations }
    procedure AddToLog(aText: String);
  public
    { Public declarations }
    procedure ExtractMasterDatabase();
    procedure ExtractTargetDatabase();
    procedure ExtractMasterScript();
    procedure ExtractTargetScript();
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

Uses DataModuleAdo, IniFiles, dbcCompObj;

function IniFileName(): String;
begin
  Result := ExtractFilePath(ParamStr(0))+'CompDemo.Ini';
end;

function AdoConnectString(WinAuth: Boolean;
                          HostName, DbName, UserName, Password: String): String;
begin
    Result := 'Provider=SQLOLEDB;' +
                        'Initial Catalog='+DbName+';';
    if HostName <> '' then Result := Result +
                        'Data Source='+HostName+';';
    if WinAuth then begin
      Result := Result + 'Integrated Security=SSPI;' +
                         'Persist Security Info=False;';
    end
    else begin
      Result := Result + 'Persist Security Info=True;';
      if UserName <> '' then
        Result := Result + 'User ID=' + UserName + ';';
      if Password <> '' then
        Result := Result + 'Password=' + Password + ';';
    end;
end;

procedure TForm1.FormCreate(Sender: TObject);
var
  MFromD, TFromD: Boolean;
begin
  EditDBMaster.Text := '';
  EditDBTarget.Text := '';
  fnScriptMaster.Text := '';
  fnScriptTarget.Text := '';
  MemoLog.Clear;
  MemoResult.Clear;
  with TIniFile.Create(IniFileName()) do begin
    EditDBMaster.Text := ReadString('Databases','MasterDBMS','');
    EditDBTarget.Text := ReadString('Databases','TargetDBMS','');
    fnScriptMaster.Text := ReadString('Databases','MasterScriptMS','');
    fnScriptTarget.Text := ReadString('Databases','TargetScriptMS','');
    MFromD := ReadBool('Databases','MasterFromDatabase',True);
    TFromD := ReadBool('Databases','TargetFromDatabase',True);
    if MFromD then rbMasterFromDatabase.Checked := True
    else rbMasterFromScript.Checked := True;
    if TFromD then rbTargetFromDatabase.Checked := True
    else rbTargetFromScript.Checked := True;
    Free;
  end;
  rbFromClick(nil);
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
  with TIniFile.Create(IniFileName()) do begin
    WriteString('Databases', 'MasterDBMS', EditDBMaster.Text);
    WriteString('Databases', 'TargetDBMS', EditDBTarget.Text);
    WriteString('Databases', 'MasterScriptMS', fnScriptMaster.Text);
    WriteString('Databases', 'TargetScriptMS', fnScriptTarget.Text);
    WriteBool('Databases','MasterFromDatabase',rbMasterFromDatabase.Checked);
    WriteBool('Databases','TargetFromDatabase',rbTargetFromDatabase.Checked);
    Free;
  end;
end;

procedure TForm1.BtnSelFileClick(Sender: TObject);
begin
  if (Sender = BtnSelDataBaseMaster) or (Sender = BtnSelDatabaseTarget) then begin
    OpenDialog1.Filter := 'All Files (*.*)|*.*';
  end;
  if (Sender = BtnSelScriptMaster) or (Sender = BtnSelScriptTarget) then begin
    OpenDialog1.Filter := 'SQL Script (*.SQL)|*.sql|SQL Script (*.DDL)|*.ddl|All Files (*.*)|*.*';
  end;
  if OpenDialog1.Execute then begin
    if Sender = BtnSelDatabaseMaster then
      EditDBMaster.Text := OpenDialog1.FileName;
    if Sender = BtnSelDatabaseTarget then
      EditDBTarget.Text := OpenDialog1.FileName;
    if Sender = BtnSelScriptMaster then
      fnScriptMaster.Text := OpenDialog1.FileName;
    if Sender = BtnSelScriptTarget then
      fnScriptTarget.Text := OpenDialog1.FileName;
  end;
end;

procedure TForm1.ExtractMasterDatabase();
var
  ErrFlg: Boolean;
begin
  with DataMdMS do begin
    ADOConnection1.Provider := 'SQLOLEDB';
    ADOConnection1.ConnectionString := AdoConnectString(rbMWin.Checked,
                                       EditHostM.Text,
                                       EditDBMaster.Text,
                                       EditMUser.Text,
                                       EditMPswd.Text);
    ErrFlg := False;
    try
      ADOConnection1.Connected := True;
    except on E: Exception do begin
      ShowMessage(E.Message);
      AddToLog('Error: '+E.Message);
      ErrFlg := True;
      ADOConnection1.Connected := False;
    end end;
    if ErrFlg then exit;
    AddToLog('Database: '+EditDBMaster.Text+' opened.');
    DBStructure1.Clear;
    MSSQLDBExtract1.ExtractDatabase;
    ADOConnection1.Connected := False;
  end;
end;

procedure TForm1.ExtractTargetDatabase;
var
  ErrFlg: Boolean;
begin
  with DataMdMS do begin
    ADOConnection2.Provider := 'SQLOLEDB';
    ADOConnection2.ConnectionString := AdoConnectString(rbTWin.Checked,
                                       EditHostT.Text,
                                       EditDBTarget.Text,
                                       EditTUser.Text,
                                       EditTPswd.Text);
    ErrFlg := False;
    try
      ADOConnection2.Connected := True;
    except on E: Exception do begin
      ShowMessage(E.Message);
      AddToLog('Error: '+E.Message);
      ErrFlg := True;
      ADOConnection2.Connected := False;
    end end;
    if ErrFlg then exit;
    AddToLog('Database: '+EditDBTarget.Text+' opened.');
    DBStructure2.Clear;
    MSSQLDBExtract2.ExtractDatabase;
    ADOConnection2.Connected := False;
  end;
end;

procedure TForm1.ExtractMasterScript();
begin
  MSSQLScriptExtract1.ScriptFileName := fnScriptMaster.Text;
  MSSQLScriptExtract1.ExtractDatabase;
end;

procedure TForm1.ExtractTargetScript();
begin
  MSSQLScriptExtract2.ScriptFileName := fnScriptTarget.Text;
  MSSQLScriptExtract2.ExtractDatabase;
end;

procedure TForm1.BtnExtractClick(Sender: TObject);
begin
  BtnExtract.Enabled := False;
  if rbTargetFromDatabase.Checked then ExtractTargetDatabase();
  if rbTargetFromScript.Checked then ExtractTargetScript();
  if rbMasterFromDatabase.Checked then ExtractMasterDatabase();
  if rbMasterFromScript.Checked then ExtractMasterScript();
  BtnExtract.Enabled := True;
  AddToLog('< Extracting finished. >');
end;

procedure TForm1.LogNextLine(Sender: TObject; LogLine: String);
begin
  AddToLog(LogLine);
end;

procedure TForm1.BtnCompareClick(Sender: TObject);
begin
  MemoResult.Lines.Clear;
  BtnCompare.Enabled := False;
  try
    DBComparer1.CompareDatabases;
    MemoResult.Lines.BeginUpdate;
    DBComparer1.SQLExec.GetScript(MemoResult.Lines);
    MemoResult.Lines.EndUpdate;
  except on E: Exception do begin
    ShowMessage(E.Message);
    MemoResult.Lines.Add('Error: '+E.Message);
  end end;
  AddToLog('< Comparing finished. >');
  BtnCompare.Enabled := True;
end;

procedure TForm1.rbFromClick(Sender: TObject);
begin
  if rbMasterFromDatabase.Checked then begin
    EditDBMaster.Enabled := True;
//    BtnSelDatabaseMaster.Enabled := True;
    fnScriptMaster.Enabled := False;
    BtnSelScriptMaster.Enabled := False;
    EditMUser.Enabled := True;
    EditMPswd.Enabled := True;
    LabelMUser.Enabled := True;
    LabelMPswd.Enabled := True;
    LabelSrvM.Enabled := True;
    EditHostM.Enabled := True;
    LabelAuthM.Enabled := True;
    rbMWin.Enabled := True;
    rbMSql.Enabled := True;
  end
  else begin
    EditDBMaster.Enabled := False;
    BtnSelDatabaseMaster.Enabled := False;
    fnScriptMaster.Enabled := True;
    BtnSelScriptMaster.Enabled := True;
    EditMUser.Enabled := False;
    EditMPswd.Enabled := False;
    LabelMUser.Enabled := False;
    LabelMPswd.Enabled := False;
    LabelSrvM.Enabled := False;
    EditHostM.Enabled := False;
    LabelAuthM.Enabled := False;
    rbMWin.Enabled := False;
    rbMSql.Enabled := False;
  end;
  if rbTargetFromDatabase.Checked then begin
    EditDBTarget.Enabled := True;
//    BtnSelDatabaseTarget.Enabled := True;
    fnScriptTarget.Enabled := False;
    BtnSelScriptTarget.Enabled := False;
    EditTUser.Enabled := True;
    EditTPswd.Enabled := True;
    LabelTUser.Enabled := True;
    LabelTPswd.Enabled := True;
    LabelSrvT.Enabled := True;
    EditHostT.Enabled := True;
    LabelAuthT.Enabled := True;
    rbTWin.Enabled := True;
    rbTSql.Enabled := True;
  end
  else begin
    EditDBTarget.Enabled := False;
    BtnSelDatabaseTarget.Enabled := False;
    fnScriptTarget.Enabled := True;
    BtnSelScriptTarget.Enabled := True;
    EditTUser.Enabled := False;
    EditTPswd.Enabled := False;
    LabelTUser.Enabled := False;
    LabelTPswd.Enabled := False;
    LabelSrvT.Enabled := False;
    EditHostT.Enabled := False;
    LabelAuthT.Enabled := False;
    rbTWin.Enabled := False;
    rbTSql.Enabled := False;
  end;
end;

procedure TForm1.ProgressUpdate(Sender: TObject; Value, MaxValue: Integer);
begin
  Gauge1.Max := MaxValue;
  Gauge1.Position := Value;
end;

procedure TForm1.BtnUpdateClick(Sender: TObject);
begin
  MSSQLExec1.ExecuteScript();
end;

procedure TForm1.AddToLog(aText: String);
begin
  MemoLog.Lines.Add(aText);
  MemoLog.Perform(EM_SCROLLCARET, 0, 0);
end;

initialization
  dbcCompObj.AddToLog := AddToLog;
end.

File: CBuilder/DBComparerDemoADO/Unit1Ado.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1Ado.h"
#include "DataModuleAdo.h"
#include <inifiles.hpp>

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "dbcClasses"
#pragma link "dbcConnection_ADO"
#pragma link "dbcCustomScriptExtract"
#pragma link "dbcDBComparer"
#pragma link "dbcDBEngine"
#pragma link "dbcDBStructure"
#pragma link "dbcMSSQLDatabaseExtract"
#pragma link "dbcMSSQLExec"
#pragma link "dbcMSSQLScriptExtract"
#pragma link "dbcSQL_Exec"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

UnicodeString __fastcall IniFileName()
{
  return ExtractFilePath(ParamStr(0)) + "CompDemo.Ini";
}
//---------------------------------------------------------------------------

UnicodeString __fastcall AdoConnectString(bool WinAuth,
  UnicodeString HostName, UnicodeString DbName, UnicodeString UserName, UnicodeString Password)
{
  UnicodeString Result = "Provider=SQLOLEDB;Initial Catalog=" + DbName + ";";
  if (HostName != "")
    Result += "Data Source=" + HostName + ";";
  if (WinAuth)
  {
    Result += "Integrated Security=SSPI;Persist Security Info=False;";
  } else
  {
    Result += "Persist Security Info=True;";
    if (UserName != "")
      Result += "User ID=" + UserName + ";";
    if (Password != "")
      Result += "Password=" + Password + ";";
  }
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  EditDbMaster->Text = "";
  EditDbTarget->Text = "";
  fnScriptMaster->Text = "";
  fnScriptTarget->Text = "";
  MemoLog->Clear();
  MemoResult->Clear();
  
  TIniFile *ini = new TIniFile(IniFileName());
  __try
  {
    EditDbMaster->Text = ini->ReadString("Databases","MasterDBMS","");
    EditDbTarget->Text = ini->ReadString("Databases","TargetDBMS","");
    fnScriptMaster->Text = ini->ReadString("Databases","MasterScriptMS","");
    fnScriptTarget->Text = ini->ReadString("Databases","TargetScriptMS","");
    bool MFromD = ini->ReadBool("Databases","MasterFromDatabase",true);
    bool TFromD = ini->ReadBool("Databases","TargetFromDatabase",true);

    if (MFromD)
      rbMasterFromDatabase->Checked = true;
    else
      rbMasterFromScript->Checked = true;
    if (TFromD)
      rbTargetFromDatabase->Checked = true;
    else
      rbTargetFromScript->Checked = true;
  }
  __finally
  {
    delete ini;
  }
  
  rbFromClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnSelFileClick(TObject *Sender)
{
  if ((Sender == BtnSelDatabaseMaster) || (Sender == BtnSelDatabaseTarget))
  {
    OpenDialog1->Filter = "All Files (*.*)|*.*";
  }
  if ((Sender == BtnSelScriptMaster) || (Sender == BtnSelScriptTarget))
  {
    OpenDialog1->Filter = "SQL Script (*.SQL)|*.sql|SQL Script (*.DDL)|*.ddl|All Files (*.*)|*.*";
  }
  if (OpenDialog1->Execute())
  {
    if (Sender == BtnSelDatabaseMaster)
      EditDbMaster->Text = OpenDialog1->FileName;
    if (Sender == BtnSelDatabaseTarget)
      EditDbTarget->Text = OpenDialog1->FileName;
    if (Sender == BtnSelScriptMaster)
      fnScriptMaster->Text = OpenDialog1->FileName;
    if (Sender == BtnSelScriptTarget)
      fnScriptTarget->Text = OpenDialog1->FileName;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnExtractClick(TObject *Sender)
{
  BtnExtract->Enabled = false;
  if (rbTargetFromDatabase->Checked)
    ExtractTargetDatabase();
  if (rbTargetFromScript->Checked)
    ExtractTargetScript();
  if (rbMasterFromDatabase->Checked)
    ExtractMasterDatabase();
  if (rbMasterFromScript->Checked)
    ExtractMasterScript();
  BtnExtract->Enabled = true;
  AddToLog("< Extracting finished. >");
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormDestroy(TObject *Sender)
{
  TIniFile *ini = new TIniFile(IniFileName());
  __try
  {
    ini->WriteString("Databases", "MasterDBMS", EditDbMaster->Text);
    ini->WriteString("Databases", "TargetDBMS", EditDbTarget->Text);
    ini->WriteString("Databases", "MasterScriptMS", fnScriptMaster->Text);
    ini->WriteString("Databases", "TargetScriptMS", fnScriptTarget->Text);
    ini->WriteBool("Databases","MasterFromDatabase",rbMasterFromDatabase->Checked);
    ini->WriteBool("Databases","TargetFromDatabase",rbTargetFromDatabase->Checked);
  }
  __finally
  {
    delete ini;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::LogNextLine(TObject *Sender, UnicodeString LogLine)
{
  AddToLog(LogLine);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnCompareClick(TObject *Sender)
{
  MemoResult->Lines->Clear();
  BtnCompare->Enabled = false;
  try
  {
    DBComparer1->CompareDatabases();
    MemoResult->Lines->BeginUpdate();
    DBComparer1->SQLExec->GetScript(MemoResult->Lines);
    MemoResult->Lines->EndUpdate();
  }
  catch (Exception &E)
  {
    Dialogs::ShowMessage(E.Message);
    MemoResult->Lines->Add("Error: " + E.Message);
  }
  AddToLog("< Comparing finished. >");
  BtnCompare->Enabled = true;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::rbFromClick(TObject *Sender)
{
  if (rbMasterFromDatabase->Checked)
  {
    EditDbMaster->Enabled = true;
//    BtnSelDatabaseMaster->Enabled = true;
    fnScriptMaster->Enabled = false;
    BtnSelScriptMaster->Enabled = false;
    EditMUser->Enabled = true;
    EditMPswd->Enabled = true;
    LabelMUser->Enabled = true;
    LabelMPswd->Enabled = true;
    LabelSrvM->Enabled = true;
    EditHostM->Enabled = true;
    LabelAuthM->Enabled = true;
    rbMWin->Enabled = true;
    rbMSql->Enabled = true;
  } else
  {
    EditDbMaster->Enabled = false;
    BtnSelDatabaseMaster->Enabled = false;
    fnScriptMaster->Enabled = true;
    BtnSelScriptMaster->Enabled = true;
    EditMUser->Enabled = false;
    EditMPswd->Enabled = false;
    LabelMUser->Enabled = false;
    LabelMPswd->Enabled = false;
    LabelSrvM->Enabled = false;
    EditHostM->Enabled = false;
    LabelAuthM->Enabled = false;
    rbMWin->Enabled = false;
    rbMSql->Enabled = false;
  }
  if (rbTargetFromDatabase->Checked)
  {
    EditDbTarget->Enabled = true;
//    BtnSelDatabaseTarget->Enabled = true;
    fnScriptTarget->Enabled = false;
    BtnSelScriptTarget->Enabled = false;
    EditTUser->Enabled = true;
    EditTPswd->Enabled = true;
    LabelTUser->Enabled = true;
    LabelTPswd->Enabled = true;
    LabelSrvT->Enabled = true;
    EditHostT->Enabled = true;
    LabelAuthT->Enabled = true;
    rbTWin->Enabled = true;
    rbTSql->Enabled = true;
  } else
  {
    EditDbTarget->Enabled = false;
    BtnSelDatabaseTarget->Enabled = false;
    fnScriptTarget->Enabled = true;
    BtnSelScriptTarget->Enabled = true;
    EditTUser->Enabled = false;
    EditTPswd->Enabled = false;
    LabelTUser->Enabled = false;
    LabelTPswd->Enabled = false;
    LabelSrvT->Enabled = false;
    EditHostT->Enabled = false;
    LabelAuthT->Enabled = false;
    rbTWin->Enabled = false;
    rbTSql->Enabled = false;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ProgressUpdate(TObject *Sender, int Value, int MaxValue)
{
  Gauge1->Max = MaxValue;
  Gauge1->Position = Value;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnUpdateClick(TObject *Sender)
{
  MSSQLExec1->ExecuteScript();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::AddToLog(UnicodeString aText)
{
  MemoLog->Lines->Add(aText);
  MemoLog->Perform(EM_SCROLLCARET, 0, 0);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractMasterDatabase()
{
  DataMdMS->ADOConnection1->Provider = "SQLOLEDB";
  DataMdMS->ADOConnection1->ConnectionString = AdoConnectString(rbMWin->Checked,
                                     EditHostM->Text,
                                     EditDbMaster->Text,
                                     EditMUser->Text,
                                     EditMPswd->Text);
  bool ErrFlg = false;
  try
  {
    DataMdMS->ADOConnection1->Connected = true;
  }
  catch (Exception &E)
  {
    Dialogs::ShowMessage(E.Message);
    AddToLog("Error: " + E.Message);
    ErrFlg = true;
    DataMdMS->ADOConnection1->Connected = false;
  }
  if (ErrFlg) return;

  AddToLog("Database: " + EditDbMaster->Text + " opened.");
  DBStructure1->Clear();
  MSSQLDBExtract1->ExtractDatabase();
  DataMdMS->ADOConnection1->Connected = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractTargetDatabase()
{
  DataMdMS->ADOConnection2->Provider = "SQLOLEDB";
  DataMdMS->ADOConnection2->ConnectionString = AdoConnectString(rbTWin->Checked,
                                     EditHostT->Text,
                                     EditDbTarget->Text,
                                     EditTUser->Text,
                                     EditTPswd->Text);
  bool ErrFlg = false;
  try
  {
    DataMdMS->ADOConnection2->Connected = true;
  }
  catch (Exception &E)
  {
    Dialogs::ShowMessage(E.Message);
    AddToLog("Error: " + E.Message);
    ErrFlg = true;
    DataMdMS->ADOConnection2->Connected = false;
  }
  if (ErrFlg) return;

  AddToLog("Database: " + EditDbTarget->Text + " opened.");
  DBStructure2->Clear();
  MSSQLDBExtract2->ExtractDatabase();
  DataMdMS->ADOConnection2->Connected = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractMasterScript()
{
  MSSQLScriptExtract1->ScriptFileName = fnScriptMaster->Text;
  MSSQLScriptExtract1->ExtractDatabase();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractTargetScript()
{
  MSSQLScriptExtract2->ScriptFileName = fnScriptTarget->Text;
  MSSQLScriptExtract2->ExtractDatabase();
}
//---------------------------------------------------------------------------


